### PR TITLE
[Core] response function interface

### PR DESCRIPTION
**Pull request by armingeiser**
This adds a top level interface for any kind of response functions. This allows to use them easily on python level e.g. for Optimization.

This interface was previously used in individual applications e.g. StructMech and ShapeOpt and would also be added to ConvectionDiffusion in #6584.

Suggested here: https://github.com/KratosMultiphysics/Kratos/pull/6584#discussion_r396311035

### CODE REVIEW

File: kratos/python_scripts/response_functions/response_function_interface.py
Comment (armingeiser) [inline kratos/python_scripts/response_functions/response_function_interface.py:1]: This is the new file in the core - Github shows it as moved...

File: applications/ShapeOptimizationApplication/tests/sensitivity_verification_process_test/run_sensitivity_verification_in_geometry_space.py
Comment (philbucher) [inline applications/ShapeOptimizationApplication/tests/sensitivity_verification_process_test/run_sensitivity_verification_in_geometry_space.py:135]: I would disable writing this in a test
It only takes time and is not used in 99% of times this test is run
and if you want to look at the output then you wouldn't delete it ;)

File: kratos/python_scripts/response_functions/response_function_interface.py
Comment (MFusseder) [inline kratos/python_scripts/response_functions/response_function_interface.py:51]: To be complete we should also add `GetConditionalGradient`.

**Review by MFusseder — COMMENTED** [review]
I am fine with the suggested modifications. Especially I appreciate the more general get gradient functions. As written I would also add something like `GetConditionalGradient`.

File: kratos/python_scripts/response_functions/response_function_interface.py
Comment (armingeiser) [reply kratos/python_scripts/response_functions/response_function_interface.py:51]: added it.

File: applications/ShapeOptimizationApplication/python_scripts/response_functions/packaging_response_base.py
Comment (dbaumgaertner) [inline applications/ShapeOptimizationApplication/python_scripts/response_functions/packaging_response_base.py:14]: Looks a bit strange to me, but I guess this is the way to go

File: kratos/python_scripts/response_functions/response_function_interface.py
Comment (dbaumgaertner) [reply kratos/python_scripts/response_functions/response_function_interface.py:1]: Just out of curiosity: Is there a reason, why you called it "Interface" rather than "Base" or just "ResponseFunction"?

File: kratos/python_scripts/response_functions/response_function_interface.py
Comment (dbaumgaertner) [inline kratos/python_scripts/response_functions/response_function_interface.py:1]: The header is missing in this file.

Comment (dbaumgaertner) [inline kratos/python_scripts/response_functions/response_function_interface.py:1]: I find it a bit unlucky, that now there are two folders called "response_functions" in the core. One under "python_scripts" and one top-level in the core for the c++ files. Not sure how to avoid this, though.

**Review by dbaumgaertner — DISMISSED** [review]
I would approve it as it looks good to me. Just some minor comments. It's good to see a unified interface of the response functions in the core. Thanks for the effort. Before merging though, it would be good to have somebody from the core guys to quickly double-check the PR @philbucher .

**Comment by dbaumgaertner** [comment]
You might also remove the trailing whitespace as suggested by codacy

**Comment by philbucher** [comment]
I would like to review this but I need a few days, I would appreciate if you wait :)

File: applications/ShapeOptimizationApplication/python_scripts/response_functions/packaging_response_base.py
Comment (armingeiser) [reply applications/ShapeOptimizationApplication/python_scripts/response_functions/packaging_response_base.py:14]: you mean the long import? Yes that's how it has to be done

File: kratos/python_scripts/response_functions/response_function_interface.py
Comment (armingeiser) [reply kratos/python_scripts/response_functions/response_function_interface.py:1]: There are no headers in the core python files, that's why i removed it. 

By calling it interface I tried to avoid the confusion with the cpp base class for adjoint responses.

Comment (armingeiser) [reply kratos/python_scripts/response_functions/response_function_interface.py:1]: Maybe also adding "interface" or similar to this folder name?

**Comment by armingeiser** [comment]
> I would like to review this but I need a few days, I would appreciate if you wait :)

No worries =)

File: kratos/python_scripts/response_functions/response_function_interface.py
Comment (dbaumgaertner) [reply kratos/python_scripts/response_functions/response_function_interface.py:1]: Well, if its something like "interface" then for sure there is only this file that fits inside such that we could pull it out anyways. In light of the strict separation between programming languages in both folders, it's maybe not too bad to leave it as is.

File: applications/ShapeOptimizationApplication/python_scripts/response_functions/packaging_response_base.py
Comment (philbucher) [reply applications/ShapeOptimizationApplication/python_scripts/response_functions/packaging_response_base.py:14]: yes this is correct / how it should be done / needs to be done
what was discussed in #5402

File: kratos/python_scripts/response_functions/response_function_interface.py
Comment (philbucher) [reply kratos/python_scripts/response_functions/response_function_interface.py:1]: I think this is ok/makes sense
Historically this is because we were not able to have submodules for a long time
Now we can I I would also use it...

File: kratos/python_scripts/response_functions/response_function_interface.py
Comment (philbucher) [reply kratos/python_scripts/response_functions/response_function_interface.py:1]: could you explain the relation btw this class and the [AdjointResponseFunction](https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/response_functions/adjoint_response_function.h)?
This should help with the naming

File: applications/ShapeOptimizationApplication/tests/sensitivity_verification_process_test/run_semi_analytic_step_size_verification.py
Comment (philbucher) [inline applications/ShapeOptimizationApplication/tests/sensitivity_verification_process_test/run_semi_analytic_step_size_verification.py:96]: idem

**Review by philbucher — COMMENTED** [review]
some comments

Just to be clear, this should also be used in Fluids?

**Comment by philbucher** [comment]
Idk why but my comments are inline with the comments of @dbaumgaertner 
=> go to `Files changed` then it is more readable

File: applications/ShapeOptimizationApplication/tests/sensitivity_verification_process_test/run_semi_analytic_step_size_verification.py
Comment (armingeiser) [reply applications/ShapeOptimizationApplication/tests/sensitivity_verification_process_test/run_semi_analytic_step_size_verification.py:96]: I see the point, however this would be something for a new MR, as this is just a whitespace change here and not related.

File: kratos/python_scripts/response_functions/response_function_interface.py
Comment (armingeiser) [reply kratos/python_scripts/response_functions/response_function_interface.py:1]: The [AdjointResponseFunction](https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/response_functions/adjoint_response_function.h) provides some of the necessary components for an adjoint sensitivity analysis:

* gradient w.r.t to the primal variables (and their derivatives)
* partial gradient w.r.t to the design variable

However it does **not** calculate the total gradient of the response function w.r.t the design variables.
This can only be obtained using the whole framework of adjoint_solver, adjoint_response_function and sensitivity_builder.

Besides the adjoint approach to calculate the gradient, there are several other ways - fore some e.g. geometrical response functions, the adjoint approach does not even make sense. 

This response function interface provides a unique interface for all possible ways to calculate the value and gradients of a response. 
This is necessary because e.g. in optimization i do not care about the implementation details (e.g. how the adjoint sensitivity analysis uses the adjoint_response_function.h ...) but i am just interested in the value and gradient of the response.

**Comment by armingeiser** [comment]
> Just to be clear, this should also be used in Fluids?
@philbucher 

It can, but i have no plans currently. The Fluid/Structural/ConvDiff use the same workflow and base classes for their adjoint sensitivity analysis - so it should be possible. 

One point that is worth mentioning is that in for structures there is sensitivity analysis only for steady state sensitivities, while for fluid it is only (?) available for transient cases.

File: kratos/python_scripts/response_functions/response_function_interface.py
Comment (philbucher) [reply kratos/python_scripts/response_functions/response_function_interface.py:1]: cool thx for the explanation

Could you please add sth like this to the docstring of the class?

**Comment by armingeiser** [comment]
@marcnunezc This is essentially the interface i gave you some time ago to test optimization with the CompressiblePotentialFlowApplication.

File: applications/ShapeOptimizationApplication/python_scripts/analyzer_internal.py
Comment (marcnunezc) [inline applications/ShapeOptimizationApplication/python_scripts/analyzer_internal.py:70]: What is the motivation to set the variable explicitly here? Isn't it more general if the variable that contains the sensitivity is not explicity set by the analyzer?

**Review by marcnunezc — COMMENTED** [review]
> 
> 
> @marcnunezc This is essentially the interface i gave you some time ago to test optimization with the CompressiblePotentialFlowApplication.

Yes! I have been using that as a custom script ever since, thanks for the effort!

I am OK with the general response interface, I only have a question regarding the reporting of the gradient.

File: applications/ShapeOptimizationApplication/python_scripts/analyzer_internal.py
Comment (armingeiser) [reply applications/ShapeOptimizationApplication/python_scripts/analyzer_internal.py:70]: The goal of this PR is to generalize the response interface, so the non generic function `GetShapeGradient` was generalized to `GetNodalGradient`. 
This now requires to specify the variable at the calling side, e.g. the optimization. The **Shape**OptimizationApplication only deals with `SHAPE` variables - and currently there are no plans to use other variables.

**Comment by philbucher** [comment]
so from my side this is ready to go in if the involved ppl are ok with it
@KratosMultiphysics/technical-committee I think we don't need to review this in detail for the same reason (we have enough more important things to do)
@marcnunezc I suggest to approve if you are ok with it

thanks for the efforts!

**Comment by philbucher** [comment]
> @armingeiser I think you didn't set your editor to auto-delete trailing whitespaces

This happens when editing files using the github online file editor =)

**Review by marcnunezc — APPROVED** [review]
Ok!

**Pull request merged**
